### PR TITLE
EDU-159: Remix Gen1 - Advanced child sub components

### PR DESCRIPTION
**Pull request by uttej-vsk**
## Description

This PR introduces remix gen 1 example for Advanced child sub-components in the custom components page in the docs. 

ticket: https://builder-io.atlassian.net/browse/EDU-324
@samijaber Please review

### CODE REVIEW

File: packages/sdks-tests/src/snippet-tests/editable-regions.spec.ts [hunk 1]
```diff
@@ -45,4 +45,19 @@ test.describe('Editable regions in custom components', () => {
     const secondText = await columns.nth(1).textContent();
     expect(secondText?.trim().toLowerCase()).toBe('column 2 text');
   });
+
+  test.describe('Remix gen1 editable regions text validation', () => {
```
Comment (samijaber) [inline packages/sdks-tests/src/snippet-tests/editable-regions.spec.ts:49]: why does remix gen1= have its own special test here? why aren't we enabling the existing tests in this spec for remix?

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      },
+    },
+  ],
+});
+
+export const loader = async ({ request }: LoaderFunctionArgs) => {
+  const page = await builder
+    .get("editable-regions", {
+      userAttributes: {
+        urlPath: `/${request.url.split("/").pop()}`,
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:37]: instead of `request.url.split("/").pop()`, we can rely on remix's file naming conventions to extract the URL path as a param. See https://github.com/BuilderIO/builder/blob/b346d8733b5b606e607c6ed5b27a2d649af08e24/packages/sdks/e2e/remix/app/routes/($slug).tsx#L44-L53

by naming the file `($slug).tsx`, remix will automatically pass the URL path value to `params.slug`. This is far cleaner as an approach, so let's do things this way for Remix moving forward

Comment (samijaber) [reply packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:37]: We do it in the snippets already doo. See: https://github.com/BuilderIO/builder/blob/b346d8733b5b606e607c6ed5b27a2d649af08e24/packages/sdks/snippets/gen1-remix/app/routes/announcements.$slug.tsx#L12-L21

File: packages/sdks/snippets/gen1-remix/app/routes/components/CustomHero.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { ReactNode } from 'react';
+
+interface CustomHeroProps {
+  children: ReactNode;
+}
+
+export const CustomHero = (props: CustomHeroProps) => {
+  return (
+    <>
+      <div>This is your component&apos;s text</div>
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/components/CustomHero.tsx:10]: `&apos;` here is not pretty. can we use a `'`?

File: packages/sdks/snippets/gen1-remix/app/routes/components/CustomColumns.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  column2: { blocks: React.ReactNode };
+  builderBlock: BuilderElement;
+}
+
+export const CustomColumns = (props: CustomColumnsProps) => {
+  return (
+    <>
+      <BuilderBlocks
+        parentElementId={props.builderBlock.id}
+        dataPath={`column1.blocks`}
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/components/CustomColumns.tsx:14]:
```suggestion
        dataPath="column1.blocks"
```

File: packages/sdks/snippets/gen1-remix/app/routes/components/CustomColumns.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        blocks={props.column1?.blocks}
+      />
+
+      <BuilderBlocks
+        parentElementId={props.builderBlock.id}
+        dataPath={`column2.blocks`}
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/components/CustomColumns.tsx:20]:
```suggestion
        dataPath="column2.blocks"
```

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+builder.init("ee9f13b4981e489a9a1209887695ef2b");
+
+Builder.registerComponent(CustomColumns, {
+  name: "MyColumns",
+  inputs: [
+    {
+      name: "column1",
+      type: "uiBlocks",
+      broadcast: true,
+      hideFromUI: true,
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:16]: i don't think we need `hideFromUI` for `uiBlocks` input types. Test that everything works as expected in the editor (that the UI block doesn't show up as an input in the Content Editor sidebase). If that's the case, let's remove it from everywhere in the snippets (this PR and elsewhere)

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx (hunk 2 above)
Comment (samijaber) [reply packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:37]: i just realized this suggestion of mine doesn't work because the routes you added in this PR are hardcoded with exact URLs 🤔

forget what i said for now, and let's stick with your approach

File: packages/sdks/snippets/gen1-remix/app/routes/custom-child.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+import { LoaderFunctionArgs } from '@remix-run/node';
+import { useLoaderData } from '@remix-run/react';
+import { CustomHero } from './components/CustomHero';
+
+builder.init('ee9f13b4981e489a9a1209887695ef2b');
+
+Builder.registerComponent(CustomHero, {
+  name: 'CustomHero',
+  inputs: [],
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/custom-child.tsx:11]: We don't need to provide an empty array, we can drop this altogether. If you can confirm that, go ahead and remove all empty `inputs: []`

File: packages/sdks/snippets/gen1-remix/app/routes/components/CustomHero.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { ReactNode } from 'react';
+
+interface CustomHeroProps {
+  children: ReactNode;
+}
+
+export const CustomHero = (props: CustomHeroProps) => {
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/components/CustomHero.tsx:7]:
```suggestion
import { ReactNode, type PropsWithChildren } from 'react';

export const CustomHero = (props: PropsWithChildren) => {
```
```

File: packages/sdks/snippets/gen1-remix/app/routes/custom-child.tsx (hunk 7 above)
Comment (uttej-vsk) [reply packages/sdks/snippets/gen1-remix/app/routes/custom-child.tsx:11]: I tried this, but it is not rendering correctly. It is expecting the inputs.

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx (hunk 6 above)
Comment (samijaber) [reply packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:16]: You resolved this comment but did not remove `hideFromUI`. can you confirm that users explicitly need to set it?

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      defaultValue: {
+        blocks: [],
+      },
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:19]: do we need the `defaultValue`s here at all? I see that the other component's `tabList` has `defaultValue: []` which is different.

ideally we can completely remove the default value to make code simpler. but if its needed, lets use the more brief version.

Comment (uttej-vsk) [reply packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:19]: @samijaber I think defaultValue here serves as a placeholder for displaying the "**+Add Blocks**" element on the UI. Without this, we could not render it on the screen.

Comment (uttej-vsk) [reply packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:19]: @samijaber I will give it a try and see if i can remove and make it work

File: packages/sdks/snippets/gen1-remix/app/routes/components/CustomColumns.tsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import { BuilderBlocks, BuilderElement } from '@builder.io/react';
+
+interface CustomColumnsProps {
+  column1: React.ReactNode;
+  column2: React.ReactNode;
+  builderBlock: BuilderElement;
+}
+
+export const CustomColumns = (props: CustomColumnsProps) => {
+  console.log('props', props);
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/components/CustomColumns.tsx:10]:
```suggestion
```

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+  console.log('page', page);
+  return { page };
+};
+
+export default function EditableRegionsPage() {
+  const { page } = useLoaderData<typeof loader>();
+
+  console.log('page', page);
+
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:42]:
```suggestion
```

File: packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+export const loader = async ({ request }: LoaderFunctionArgs) => {
+  const page = await builder
+    .get('editable-regions', {
+      userAttributes: {
+        urlPath: `/${request.url.split('/').pop()}`,
+      },
+    })
+    .toPromise();
+
+  console.log('page', page);
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/editable-region.tsx:34]:
```suggestion
```

File: packages/sdks-tests/src/snippet-tests/editable-regions.spec.ts [hunk 13]
```diff
@@ -22,27 +22,17 @@ test.describe('Editable regions in custom components', () => {
     page,
     packageName,
   }) => {
-    test.skip(!['react', 'angular', 'angular-ssr'].includes(packageName));
+    test.skip(
+      !['react', 'angular', 'angular-ssr', 'gen1-react', 'gen1-remix'].includes(packageName)
```
Comment (samijaber) [inline packages/sdks-tests/src/snippet-tests/editable-regions.spec.ts:26]: this test is failing for `gen1-react`: https://github.com/BuilderIO/builder/actions/runs/12189792080/job/34005639784?pr=3750#step:5:1220

i see you added `'gen1-react'` here, but you didn't add any code to support it in this PR. is this a mistake?

File: packages/sdks-tests/src/snippet-tests/advanced-child.spec.ts [hunk 14]
```diff
@@ -25,25 +25,30 @@ test.describe('Advanced child sub components', () => {
...
-      'component.options.tabList.1.blocks',
-      'component.options.tabList.0.blocks'
-    );
+    const tab1Content = await page.waitForSelector('div:has-text("Tab 1 Content")', {
+      state: 'visible',
+    });
+    expect(await tab1Content.isVisible()).toBe(true);
+
+    const tab2ContentHidden = await page.$('div:has-text("Tab 2 content")');
+    expect((await tab2ContentHidden?.isVisible()) ?? false).toBe(false);
```
Comment (samijaber) [inline packages/sdks-tests/src/snippet-tests/advanced-child.spec.ts:43]: `expect((await element?.isVisible()) ?? false).toBe(false)` is confusing and hard to follow. Also, using `page.$` is not recommended, we should rely on `page.locator` whenever possible.

based on https://github.com/microsoft/playwright/issues/18815#issuecomment-1315751818

you should be able to do:

```suggestion
    expect(await page.locator('div').filter({ hasText: 'Tab 2 content' }).isVisible()).toBeFalsy();
```

File: packages/sdks-tests/src/snippet-tests/advanced-child.spec.ts [hunk 15]
```diff
@@ -25,25 +25,30 @@ test.describe('Advanced child sub components', () => {
...
+
+    await page.click('button:has-text("Tab 2")');
+    const tab2Content = await page.waitForSelector('div:has-text("Tab 2 content")', {
+      state: 'visible',
+    });
+    expect(await tab2Content.isVisible()).toBe(true);
+
+    const tab1ContentHidden = await page.$('div:has-text("Tab 1 Content")');
+    expect((await tab1ContentHidden?.isVisible()) ?? false).toBe(false);
```
Comment (samijaber) [inline packages/sdks-tests/src/snippet-tests/advanced-child.spec.ts:52]: same here

File: packages/sdks-tests/src/snippet-tests/advanced-child.spec.ts (hunk 14 above)
Comment (uttej-vsk) [reply packages/sdks-tests/src/snippet-tests/advanced-child.spec.ts:43]: @samijaber I have made it much simpler in the test case by calling a helper function. Let me know if that is okay.

File: packages/sdks-tests/src/snippet-tests/editable-regions.spec.ts (hunk 13 above)
Comment (uttej-vsk) [reply packages/sdks-tests/src/snippet-tests/editable-regions.spec.ts:26]: @samijaber Oh yeah, my mistake. Removing it.

File: packages/sdks/snippets/gen1-remix/app/routes/advanced-child.tsx [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+      type: 'array',
+      subFields: [
+        {
+          name: 'tabName',
+          type: 'string',
+        },
+        {
+          name: 'blocks',
+          type: 'uiBlocks',
+          hideFromUI: true,
```
Comment (samijaber) [inline packages/sdks/snippets/gen1-remix/app/routes/advanced-child.tsx:23]:
```suggestion
```

**Review by samijaber — APPROVED** [review]
LGTM, just remove that one unnecessary field
